### PR TITLE
[blocked by #770] Cleaning subsequent calls in Trainer [wip]

### LINKED ISSUE #633: Trainers' .fit() mimics .test() after first call to .test() + .test() doesn't print metrics
## 🐛 Bug

1) After first call to `Trainer.test()` all subsequent calls to `Trainer.fit()` exhibit output behavior of `Trainer.test()`
2) `Trainer.test()` doesn't print metrics (and returns `None`) returned by `LightningModule.test_end()`

### To Reproduce
Run following code in a `Python 3.6.8` env with `torch=1.3.1` and `pytorch_lightning=0.5.3.2` installed:
#### Code sample

<details>
  <summary>Click to view the code sample.</summary>

```python
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import TensorDataset, DataLoader

import pytorch_lightning as pl

print(torch.__version__, pl.__version__)


class TestModule(pl.LightningModule):

    def __init__(self, bs):
        super(TestModule, self).__init__()
        self.fc = nn.Linear(2, 2)
        self.bs = bs
        self.criterion = nn.MSELoss()

    def forward(self, x):
        x = self.fc(x)
        return x
    
    def training_step(self, batch, batch_nb):
        x, y = batch
        y_hat = self.forward(x)
        return {'loss': self.criterion(y_hat, y)}

    def test_step(self, batch, batch_nb):
        x, y = batch
        y_hat = self.forward(x)
        return {'test_loss': self.criterion(y_hat, y)}
    
    def test_end(self, outputs):
        test_loss = torch.stack([x['test_loss'] for x in outputs]).mean()
        test_metrics = {'test_loss': test_loss}
        return {'progress_bar': test_metrics, 'log': test_metrics}

    def configure_optimizers(self):
        self.optimizer = optim.Adam(self.parameters())
        return self.optimizer

    @pl.data_loader
    def train_dataloader(self):
        x = torch.rand(1000, 2) - 0.5
        y = torch.sign(x)
        ds = TensorDataset(x, y)
        dl = DataLoader(ds, batch_size=self.bs, shuffle=True)
        return dl

    @pl.data_loader
    def test_dataloader(self):
        x = torch.rand(100, 2) - 0.5
        y = torch.sign(x)
        ds = TensorDataset(x, y)
        dl = DataLoader(ds, batch_size=self.bs * 2)
        return dl


device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

net = TestModule(bs=32).to(device)

epochs = 10
trainer = pl.Trainer(gpus=-1, max_nb_epochs=epochs, min_nb_epochs=epochs)

trainer.fit(net)
trainer.test()
trainer.fit(net)
trainer.fit(net)
```

</details>

#### Code output
Output of the sequence of calls `fit` -> `test` -> `fit` -> `fit`
```
1.3.1 0.5.3.2
Epoch 10: 100%|██████████| 32/32 [00:00<00:00, 272.26batch/s,
batch_nb=31, gpu=0, loss=1.009, v_nb=65]
Testing: 100%|██████████| 2/2 [00:00<00:00, 222.23batch/s]
Testing: 100%|██████████| 2/2 [00:00<00:00, 357.40batch/s]
Testing: 100%|██████████| 2/2 [00:00<00:00, 380.52batch/s]
```
### Expected behavior

1) `Trainer.fit()` should always run model training, even after it was already tested once via `Trainer.test()`.
2) `Trainer.test()` should return metrics produced by `LightningModule.test_end()`. What is processing the whole test dataset through the model good for if not collecting performance metrics?

### Environment

<details>
  <summary>Click to view the environment.</summary>

```
PyTorch version: 1.3.1 [0/47800]
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: TITAN V
Nvidia driver version: 418.87.00
cuDNN version: Probably one of the following:
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn.so.7
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7
/usr/local/cuda-9.0/targets/x86_64-linux/lib/libcudnn.so.7

Versions of relevant libraries:
[pip3] botorch==0.1.4
[pip3] gpytorch==0.3.6
[pip3] numpy==1.17.4
[pip3] pytorch-lightning==0.5.3.2
[pip3] torch==1.3.1
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.4.2
[conda] Could not collect
```

</details>

### PR DESCRIPTION
**Pull request by abdullah-alnahas**
This PR does the following:
- After the first call to `Trainer.test()` all subsequent calls to `Trainer.fit()` will NOT exhibit output behavior of `Trainer.test()`, 
- Both `Trainer.test()` and `Trainer.test(model)` WILL return the metrics returned by `LightningModule.test_end()`.

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes #633. Get Trainer.test() to return the metrics returned by LightningModule.test_end() and get all subsequent calls to Trainer.fit() to work properly

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
I definitely had fun 🙃.

### CODE REVIEW

**Comment by Borda** [comment]
Pls name the PR properly and if it is work in progress add "[wip]" to the end of the name...

**Comment by abdullah-alnahas** [comment]
Thank you for your response. I am new to GitHub, so sorry for troubling you.
I have renamed the PR, is it good now?
Why Travis is failing?
BTW, I think I have solved the bugs mentioned by #633 .

File: pytorch_lightning/trainer/evaluation_loop.py [hunk 1]
```diff
@@ -334,6 +335,10 @@ def run_evaluation(self, test=False):
             self.checkpoint_callback.on_epoch_end(epoch=self.current_epoch,
                                                   logs=self.callback_metrics)
 
+        # return evaluation results called from Trainer.test()
+        if test:
```
Comment (Borda) [inline pytorch_lightning/trainer/evaluation_loop.py]: why do not return it always?

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -908,6 +911,9 @@ def test(self, model=None):
         """
         self.testing = True
         if model is not None:
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]: there is no need for this if/else split, just call `eval_results = self.run_evaluation(test=True, model=model)` and if the model is none that it does the same as default parameter...

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -908,6 +911,9 @@ def test(self, model=None):
...
-            self.fit(model)
+            eval_results = self.run_evaluation(test=True, model=model)
         else:
-            self.run_evaluation(test=True)
+            eval_results = self.run_evaluation(test=True)
+        self.testing = False
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]: not sure if this is needed...

File: pytorch_lightning/trainer/evaluation_loop.py (hunk 1 above)
Comment (abdullah-alnahas) [reply pytorch_lightning/trainer/evaluation_loop.py]: I guess this function could be called for both validation and test evaluations, and in issue #633 it was only required for Trainer.test() to return the evaluation results. However, if called for validation evaluation, it should be logged not returned, as far as I know.

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (abdullah-alnahas) [reply pytorch_lightning/trainer/trainer.py]: Thanks. I will fix it.

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (abdullah-alnahas) [reply pytorch_lightning/trainer/trainer.py]: Without setting `self.testing=False` any subsequent call to `Trainer.fit()` will act as `Trainer.test()`.

**Comment by abdullah-alnahas** [comment]
@Borda Could you have a look? I think my PR solves issue #633

**Comment by williamFalcon** [comment]
@abdullah-alnahas awesome. Mind writing a test for this? We can merge once that's done.

**Comment by williamFalcon** [comment]
@abdullah-alnahas would be great to merge this! 
New release coming in a few days :) 
Would love to get this in there

**Comment by abdullah-alnahas** [comment]
@williamFalcon Thanks for your support! I have committed the tests. Please let me know what else is needed when you get the chance.

**Comment by abdullah-alnahas** [comment]
@williamFalcon when I only run test_trainer.py my new added test, i.e. `test_test_method` passes, however, CI fails. Any idea how to fix that?

File: pytorch_lightning/trainer/evaluation_loop.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/trainer/evaluation_loop.py]: then pls return it always :]

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (Borda) [reply pytorch_lightning/trainer/trainer.py]: per comment above, if you return it always, you do not need to swap it there and back...

File: tests/test_trainer.py [hunk 4]
```diff
@@ -411,5 +412,54 @@ class CurrentTestModel(
     trainer.test()
 
 
+def test_test_method(tmpdir):
```
Comment (Borda) [inline tests/test_trainer.py]: rather `test_trainer_test_method`

File: tests/test_trainer.py [hunk 5]
```diff
@@ -411,5 +412,54 @@ class CurrentTestModel(
...
+    """Verify Trainer.test() is working properly."""
+
+    tutils.reset_seed()
+
+    class CurrentTestModel(
```
Comment (Borda) [inline tests/test_trainer.py]: `LocalTestModel` ?

File: tests/test_trainer.py [hunk 6]
```diff
@@ -411,5 +412,54 @@ class CurrentTestModel(
...
+
+    # fit model
+    trainer = Trainer(**trainer_options)
+    trainer.fit(model)
+
+    # run the test method
+    test_result = trainer.test()
+
+    # verify test method returns what is returned by test_end
+    assert 'progress_bar' in test_result and 'test_acc' \
```
Comment (Borda) [inline tests/test_trainer.py]: I would rewrite is as separate asserts... so it is cleaner to read...

File: tests/test_trainer.py (hunk 5 above)
Comment (abdullah-alnahas) [reply tests/test_trainer.py]: Excuse me, I don't understand.

Comment (Borda) [reply tests/test_trainer.py]: It was about remane, but ignore it...

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (abdullah-alnahas) [reply pytorch_lightning/trainer/trainer.py]: @Borda Thx a lot for your time. I return it always now. However, I still not understand why I shouldn't change `self.testing` back and forth.

Comment (Borda) [reply pytorch_lightning/trainer/trainer.py]: Sure, pls check #770

Comment (abdullah-alnahas) [reply pytorch_lightning/trainer/trainer.py]: Changes proposed by #770 are really good. I guess when you merge #770 to master, current PR will be irrelevant!

Comment (Borda) [reply pytorch_lightning/trainer/trainer.py]: cool! let's wait it is in master and stay tune in meantime ;]

**Comment by williamFalcon** [comment]
@Borda @abdullah-alnahas let's merge this?

**Comment by Borda** [comment]
This one waiting on enum modes test/val/etc. introduced in #770 so no much to do until the other is merged... :]

**Comment by williamFalcon** [comment]
fixed in 0.7.1!

**Pull request closed without merging**